### DISCUSSION TITLE
Internationalizing numerals

### DISCUSSION THREAD

**Discussion by eddieantonio** [hedyorg/hedy#1920] [Language Ideas]
This is a discussion about how Hedy can support different numeral systems, especially for languages that do not use Western Arabic numerals (i.e., the digits 0123456789).

Hedy currently has two burgeoning translations that may want to use their own numerals: Arabic and Bengali.

Parsing numerals should be fairly straightforward. Around level 6, calculations are now possible. That means digits are now printed to the screen. At this point, we might want to parse the ASCII digits, and then produce numerals in the more appropriate numbering system.

My question is: **is this desirable?** I don't know Arabic or Bengali—would speakers of these languages prefer or benefit from these changes?

---

I just tried this, and **Hedy (at level 1) already supports parsing (Eastern) Arabic and Bengali numbers!** 🤯 Try this in [Turtle](https://www.hedycode.com/hedy#turtle):

```hedy
forward ৫০
turn right
forward ৫০
turn right
forward ৫০
turn right
forward ৫০
``` 

At level 6, things start to break down. I believe `INT` is the wrong terminal here: https://github.com/Felienne/hedy/blob/0a81750a0301431d38ed5d6ba6073f7d7300514e/grammars/level6-Additions.lark#L25

**Comment by Felienne** [comment]
Thanks for pushing this @eddieantonio!! I think we should support it, I mean we are going all the way now with doing it proper right?!

@alaaeddin-swidan what do you think about (Eastern) Arabic numbers? I happen to have a call later today with the Hindi/Bengali translators, I will ask them, but I know that in an earlier version of their translation they did use their own numbers which maks sense of course and I manually changed them

**Comment by Felienne** [comment]
> I just tried this, and **Hedy (at level 1) already supports parsing (Eastern) Arabic and Bengali numbers!** 🤯 Try this in [Turtle](https://www.hedycode.com/hedy#turtle):
> 
> ```
> forward ৫০
> turn right
> forward ৫০
> turn right
> forward ৫০
> turn right
> forward ৫০
> ```

Yeah we allow texts to be of any letter type so that should work. I think it would not be superhard to include to be honest, just replacing the lark builtin `INT` here by a broader set of characters that we know are numbers, but of course I am overlooking simple corner cases for lack of knowing any other numeral type haha.

**Comment by alaaeddin-swidan** [comment]
Another good point @eddieantonio!

- Related to the main question, I think it is not about preference but more about the localization rules that are applied when using Hedy. The idea is to keep the users within their comfort zone, without switching between keyboards to get things done. In the case of numbers, when I write Arabic in my machine the numbers will automatically be written in the Arabic letters (0123...9). But I suspect this is the case because my machine is set up in this way (primarily the OS (Windows) language being English). People living in the middle east will have different/various setups in their machines which will make their default number system to be sometimes the Eastern Arabic (٠١٢٣٤٥٦٧٨٩). Short answer: i think we need to support all/some of the number systems below for better experience.   
![2022-02-08_11h17_04](https://user-images.githubusercontent.com/7933089/152969753-46b3d062-26c7-4228-9f61-003449a5edb6.png)
 
-   Interesting finding that of the turtle @Felienne! I tried it with the eastern Arabic numbers (above) and it works fine. But the same numbers do not work with the mathematical operators in level 6 for example. See screenshots: 

![2022-02-08_11h34_16](https://user-images.githubusercontent.com/7933089/152970515-160de9ed-223f-446e-92ef-a09929e32ccf.png)

![2022-02-08_11h36_20](https://user-images.githubusercontent.com/7933089/152970510-e6b66717-fd10-4719-91db-87568b245db0.png)

**Reply by Felienne** [reply]
> * Interesting finding that of the turtle @Felienne! I tried it with the eastern Arabic numbers (above) and it works fine. 

ok this is embarrassing but I have no clue *why* it works?! I assume that Skulpt/Python can already interpret these numbers and convert them to their western counterparts? 

In level 6 it fails because of the parser, but if we allow it in the grammar maybe it will just work? I will try to make a draft for this.

**Comment by Felienne** [comment]
Well what do you know?!

Python does not allow to use Hindi numerals (in fact @eddieantonio, we had an issue open for this already!! https://github.com/Felienne/hedy/issues/1043)

which describes that it does not parse:

![image](https://user-images.githubusercontent.com/1003685/152995610-280b9de7-8fc7-4961-b14c-2f49f8aebe31.png)

Buuuuuuuuut (thank you Python developers!!!) you *can* convert strings in non-Latin numbers and that works, which is how our turtle (accidentally) works:

![image](https://user-images.githubusercontent.com/1003685/152995430-e52e7a53-e340-426e-a749-4ef0d1d50c45.png)

So I think I can make this work with a little bit of parser changes and a lot of casts!
